### PR TITLE
add inversify-cpp/2.1.0

### PR DESCRIPTION
**Pull request by mosure**
Specify library name and version:  **inversify-cpp/2.1.0**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

New addition to ConanCenter.
Standalone library.
I am the author.

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by mosure** [comment]
@conan-center-bot remove label

File: recipes/inversify-cpp/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+from conans import ConanFile, tools
+import os
+
+required_conan_version = ">=1.33.0"
+
+
+class InversifyCppConan(ConanFile):
+    name = "inversify-cpp"
+    description = "C++17 inversion of control and dependency injection container library"
+    topics = ("conan", "inversify-cpp", "ioc", "container", "dependency", "injection")
```
Comment (uilianries) [inline recipes/inversify-cpp/all/conanfile.py:10]:
```suggestion
    topics = ("ioc", "container", "dependency", "injection")
```

File: recipes/inversify-cpp/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/mosure/inversify-cpp"
+    license = "MIT"
+    settings = "compiler"
+    no_copy_source = True
+    
+    def validate(self):
+        tools.check_min_cppstd(self, "17")
```
Comment (uilianries) [inline recipes/inversify-cpp/all/conanfile.py:18]:
```suggestion
    @property
    def _compilers_minimum_version(self):
        return {
            "gcc": "9",
            "Visual Studio": "15.7",
            "clang": "7",
            "apple-clang": "11",
        }

    def validate(self):
        if self.settings.compiler.cppstd:
            tools.check_min_cppstd(self, 17)
        minimum_version = self._compilers_minimum_version.get(str(self.settings.compiler), False)
        if minimum_version:
            if tools.Version(self.settings.compiler.version) < minimum_version:
                raise ConanInvalidConfiguration("{} requires C++17, which your compiler does not support.".format(self.name))
        else:
            self.output.warn("{} requires C++17. Your compiler is unknown. Assuming it supports C++17.".format(self.name))
```

Need to validate compiler versions

File: recipes/inversify-cpp/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    def package(self):
+        self.copy(pattern="LICENSE", dst="licenses", src=self._source_subfolder)
+        self.copy(pattern="*", dst="include", src=os.path.join(self._source_subfolder, "include"))
+
+    def package_id(self):
+        self.info.header_only()
+
+    def package_info(self):
+        self.cpp_info.includedirs = ['include']
```
Comment (uilianries) [inline recipes/inversify-cpp/all/conanfile.py:35]:
```suggestion
```

Not needed.

File: recipes/inversify-cpp/all/test_package/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CONAN_DISABLE_CHECK_COMPILER ON)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_executable(${PROJECT_NAME} test_package.cpp)
+target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
```
Comment (uilianries) [inline recipes/inversify-cpp/all/test_package/CMakeLists.txt:12]:
```suggestion
cmake_minimum_required(VERSION 3.1)
project(test_package CXX)

include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
conan_basic_setup(TARGETS)

find_package(inversify-cpp CONFIG REQUIRED)

add_executable(${PROJECT_NAME} test_package.cpp)
target_link_libraries(${PROJECT_NAME} inversify-cpp::inversify-cpp)
```

File: recipes/inversify-cpp/all/test_package/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+from conans import ConanFile, CMake, tools
+import os
+
+
+class TestPackageConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
```
Comment (uilianries) [inline recipes/inversify-cpp/all/test_package/conanfile.py:7]:
```suggestion
    generators = "cmake", "cmake_find_package_multi"
```

File: recipes/inversify-cpp/all/test_package/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    
+    def validate(self):
+        tools.check_min_cppstd(self, "17")
+
```
Comment (uilianries) [inline recipes/inversify-cpp/all/test_package/conanfile.py:11]:
```suggestion
```

do not validate it on test package.

File: recipes/inversify-cpp/all/test_package/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        if not tools.cross_building(self.settings):
```
Comment (uilianries) [inline recipes/inversify-cpp/all/test_package/conanfile.py:18]:
```suggestion
        if not tools.cross_building(self):
```

self.settings does not work when using build context.

**Review by uilianries — COMMENTED** [review]
Thank you for contribution, please take a look on my review.

**Review by uilianries — DISMISSED** [review]
LGTM

**Review by SSE4 — DISMISSED** [review]

File: recipes/inversify-cpp/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    def _compilers_minimum_version(self):
+        return {
+            "gcc": "7",
+            "Visual Studio": "15.7",
+            "clang": "6",
+            "apple-clang": "11",
+        }
+
+    def validate(self):
+        tools.check_min_cppstd(self, 17)
```
Comment (prince-chrismc) [inline recipes/inversify-cpp/all/conanfile.py:28]:
```suggestion
        if self.settings.compiler.get_safe("cppstd"):
            tools.check_min_cppstd(self, 17)
```

**Review by prince-chrismc — CHANGES_REQUESTED** [review]
One tiny issue https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#why-is-a-toolscheck_min_cppstd-call-not-enough

**Comment by mosure** [comment]
Hello, I am reviewing [args-parser/all/test_package/conanfile.py](https://github.com/conan-io/conan-center-index/blob/master/recipes/args-parser/all/test_package/conanfile.py). Given [`args-parser`'s `cppstd` validation](https://github.com/conan-io/conan-center-index/blob/master/recipes/args-parser/all/conanfile.py#L30), I would expect `inversify-cpp`'s test package to behave similarly, however, the `test_package` does not seem to be running with correct `cppstd >= 17` after incorporating `cppstd` guard.

**Comment by prince-chrismc** [comment]
ahh excellent pitfal... You are missing this little bit of magic

https://github.com/conan-io/conan-center-index/blob/e6b59d950aac0909e777f962d1fda9ab428c1995/recipes/args-parser/all/test_package/CMakeLists.txt#L15-L18

***

So with the change (adding the `if` statement) we are never hitting the "check min" because CCI infra currently does not handle the `cppstd` setting. Previously it was using the compiler default (and block compilers that support it with that setting). So now we need to make sure that is set in the test_package in the cmake

**Comment by prince-chrismc** [comment]
You can block windows MT since it's the only failure

File: recipes/inversify-cpp/all/conanfile.py [hunk 9]
```diff
@@ -37,7 +37,7 @@ def validate(self):
 
     @property
     def _source_subfolder(self):
-        return "source_subfolder"
+        return os.path.join(self.source_folder, "source_subfolder")
```
Comment (prince-chrismc) [inline recipes/inversify-cpp/all/conanfile.py:40]: This should not be required (the default relative path always starts from here

File: recipes/inversify-cpp/all/conanfile.py [hunk 10]
```diff
@@ -48,3 +48,8 @@ def package(self):
 
     def package_id(self):
         self.info.header_only()
+
+    def package_info(self):
+        self.cpp_info.names["cmake_find_package"] = "inversify-cpp"
+        self.cpp_info.names["cmake_find_package_multi"] = "inversify-cpp"
+        self.cpp_info.names["pkg_config"] = "inversify-cpp"
```
Comment (prince-chrismc) [inline recipes/inversify-cpp/all/conanfile.py:55]: I did not find these defined in your project + they are the default from Conan
We only explicit set the ones defined in the project. 

I think these can be removed

**Review by prince-chrismc — APPROVED** [review]

**Review by ericLemanissier — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Pull request merged**